src/snowflake/snowpark/mock/_snowflake_data_type.py

```diff
@@ -234,6 +242,74 @@ def calculate_type(c1: ColumnType, c2: Optional[ColumnType], op: Union[str]):
     )
 
 
+def coerce_t1_into_t2(t1: DataType, t2: DataType) -> Optional[DataType]:
+    if t1 == t2:
+        return t2
+    elif isinstance(t1, NullType):
+        return t2
+    if isinstance(t1, StringType):
+        if isinstance(t2, StringType):
+            l1 = t1.length or StringType._MAX_LENGTH
+            l2 = t2.length or StringType._MAX_LENGTH
+            return StringType(max(l1, l2))
+        elif isinstance(
+            t2,
+            (
+                _FractionalType,
+                _IntegralType,
+                DateType,
+                TimeType,
+                TimestampType,
+                VariantType,
+            ),
+        ):
+            return t2
+    elif isinstance(t1, _IntegralType):
+        if isinstance(t2, _IntegralType):
+            res = calculate_type(ColumnType(t1, True), ColumnType(t2, True), "+")
+            return res.datatype
+        elif isinstance(t2, (_FractionalType, VariantType, BooleanType)):
+            return t2
+    elif isinstance(t1, _FractionalType):
+        if isinstance(t2, (_FractionalType, BooleanType, VariantType)):
+            return t2
+    elif isinstance(t1, BooleanType):
+        if isinstance(t2, (StringType, VariantType)):
+            return t2
+    elif isinstance(t1, DateType):
+        if isinstance(t2, (TimestampType, VariantType)):
+            return t2
+    elif isinstance(t1, ArrayType):
+        if isinstance(t2, ArrayType):
+            if t1.element_type == t2.element_type:
+                return t2
+            else:
+                return ArrayType(VariantType())
+        elif isinstance(t2, VariantType):
+            return t2
+    elif isinstance(t1, MapType):
+        if isinstance(t2, MapType):
+            if t1.key_type == t2.key_type and t2.value_type == t1.value_type:
+                return t2
+            else:
+                return MapType(key_type=VariantType(), value_type=VariantType())
+        elif isinstance(t2, VariantType):
+            return t2
+    elif isinstance(t1, (TimeType, TimestampType, MapType, ArrayType)):
```

Comment:
Logically I think TimeType -> TimestampType and DateType -> TimestampType should be valid as well, but I'm not sure if that holds up in practice.

Reply:
`DateType -> TimestampType` is supported on line 284, `TimeType -> TimestampType` is not supported by my tests.